### ISSUE TITLE
Add Muskingum-Cunge Routing

### ISSUE THREAD

**Issue by micahjohnson150** [USDA-ARS-NWRC/swiflow#6]
* swiflow version: 0.3.0

### Description

For more realistic routing, @etrujil has suggested that we should have a muskingum Cunge routing method implemented.

**Comment by micahjohnson150** [comment]
Looking at existing solutions.

Found [Hapi](https://github.com/MAfarrag/HAPI/blob/master/Hapi/Routing.py). They have a muskingum implemented but when I tried to use it they require gdal the python library. I am not quite ready to make that commitment for SWIFlow but its an option.

**Comment by micahjohnson150** [comment]
@etrujil  I think I am finished with this. I will do some more testing tomorrow to be totally sure. But before closing I would like to look over with you and just double check the theory. 

Immediately I have only one question and it may be for Adrien but in case you know that answer:

Q_r (Representative flowrate) can be zero sometimes. Since it is in the denominator of an equation what should we do? Currently I am usin 1e-6 as a substitute for zero to get pass errors.

**Comment by micahjohnson150** [comment]
@Hedrick-ARS , @scotthavens , @etrujil , @robertson-mark 

I would like some review of the equations from the [Galiice et al 2016](https://www.geosci-model-dev.net/9/4491/2016/gmd-9-4491-2016.pdf) paper in section 2.2.1 which have been implemented in the route() loop in routing_method.py under MuskingumCunge class starting at line 152: .https://github.com/USDA-ARS-NWRC/swiflow/blob/3cb8cb5c8c64502fbdc67c26be909ea654e23df2/swiflow/routing_models.py#L152

and ending at line 231
https://github.com/USDA-ARS-NWRC/swiflow/blob/3cb8cb5c8c64502fbdc67c26be909ea654e23df2/swiflow/routing_models.py#L231
 
The code should be pretty close to whats in the paper. I tried to match the variables as best as possible. I have poured over this the last couple days looking for a typo as my results are not very good. I suspect there is something small somewhere. Please feel free to ask any questions.

**Comment by scotthavens** [comment]
Is there a test that we can run to look at what the `route()` function is actually doing?

**Comment by micahjohnson150** [comment]
I have just been using the tollgate example and changing timeframes in the tollgate example.
```
cd ./examples/tollgate
swiflow config.ini
```

**Comment by scotthavens** [comment]
All the equations appear to be the same as the what is in the paper. I'd recommend adding the equation numbers in the code to help anyone following allong.

What is the exact problem that you are seeing? What do you mean by the results aren't very good?

**Comment by scotthavens** [comment]
A few things I noticed which are pushed to the `scott_review` branch.

- `Q_df` was not numeric but an object type
- `Q_r` is the representative flow rate (mean flow rate) from the upstream and current cell. If there is no flow in anything, then by default `Q_ti` will be 0 because all the flow rates in equation 2 are 0. Then you don't have to set `Q_r` to a small number which won't make most of the data frame be NaN. Since there is no flow until 10/8, the output from swiflow should also be 0.
- The first time step with input SWI is 10/8 which is the first timestep that the streamflow becomes negative for some of the reaches. The reach `1` is negative because `C3` is negative or the timestep is to large for `x_i` and `l_i`. So it looks like the first timestep is numerically instable and we'll have to use equation 8 to set `dt`.

I added the tollgate example to the tests, it was an easier way to debug swiflow than through the examples. Other than that, all the equations are good, just need to implement the instability part.
